### PR TITLE
Move get_swap_chain_preferred_format from device to adapter and check formats

### PR DESCRIPTION
**Pull request by niklaskorz**
**Connections**

https://github.com/gfx-rs/wgpu-rs/issues/123#issuecomment-757831458

**Description**

Not all devices actually support `Bgra8UnormSrgb`. If this is the case, a similar format like `Rgba8UnormSrgb` can be used.

**Testing**

https://github.com/niklaskorz/wgpu/blob/e5ebfa70a4416c1b579d32566b5c1c8190dac189/learn-wgpu/src/lib.rs#L137

- Output on MacBook Pro 13" Intel 2020 macOS Metal: `Format: Bgra8UnormSrgb`
- Output on OnePlus 7 Pro Android Vulkan: `Format: Rgba8UnormSrgb`

### CODE REVIEW

File: wgpu-core/src/device/mod.rs [hunk 1]
```diff
@@ -3887,12 +3897,46 @@ impl<G: GlobalIdentityHandlerFactory> Global<G> {
...
+            }
+            // If not available, use Rgba8UnormSrgb
+            if formats.contains(&conv::map_texture_format(TextureFormat::Rgba8UnormSrgb, device.private_features)) {
+                return Ok(TextureFormat::Rgba8UnormSrgb);
+            }
+            // TODO: are there devices that support neither of these?
+        }
+
+        // If no formats were returned, use Bgra8UnormSrgb
         Ok(TextureFormat::Bgra8UnormSrgb)
```
Comment (kvark) [inline wgpu-core/src/device/mod.rs:3940]: we'd also hit this path if the formats are returned, but neither of the branches was taken

File: wgpu-core/src/device/mod.rs [hunk 2]
```diff
@@ -3887,12 +3897,46 @@ impl<G: GlobalIdentityHandlerFactory> Global<G> {
 
     pub fn device_get_swap_chain_preferred_format<B: GfxBackend>(
```
[the remark this answers is not in the document]
Comment (niklaskorz) [reply wgpu-core/src/device/mod.rs:3898]: Sure!

File: wgpu-core/src/device/mod.rs (hunk 1 above)
Comment (niklaskorz) [reply wgpu-core/src/device/mod.rs:3940]: Do you think it'd be better to return an error in this case? Or just take the first format in the list, even if it's a single-color format?

Comment (kvark) [reply wgpu-core/src/device/mod.rs:3940]: let's take the first format, still better than an error :)

Comment (niklaskorz) [reply wgpu-core/src/device/mod.rs:3940]: I had another look at the definition of `getSwapChainPreferredFormat(adapter)` in the WebGPU spec, where it says:

> Return an optimal GPUTextureFormat to use when creating a GPUSwapChain with the given adapter. Must be one of the supported swap chain formats.
> The supported swap chain formats are a set of GPUTextureFormats that must be supported when specified as a GPUSwapChainDescriptor.format regardless of the given GPUSwapChainDescriptor.device, initially set to: «"bgra8unorm", "bgra8unorm-srgb", "rgba8unorm", "rgba8unorm-srgb"».

So I guess it is sufficient to check for these four, and then fallback to the first format in the list, in the unlikely case that none of these is supported.

Comment (niklaskorz) [reply wgpu-core/src/device/mod.rs:3940]: On the other hand, there are more formats described in HAL than there are texture formats defined in wgpu. So for now, I think checking these four mentioned in the spec and returning an error instead might be the easier choice until someone comes up with a use case where these four don't work.

File: wgpu-core/src/instance.rs [hunk 3]
```diff
@@ -250,6 +250,44 @@ impl<B: GfxBackend> Adapter<B> {
         }
     }
 
+    pub fn get_swap_chain_preferred_format(
+        &self,
+        surface: &mut Surface,
```
Comment (kvark) [inline wgpu-core/src/instance.rs:255]: does this need to be mutable?

File: wgpu-core/src/instance.rs [hunk 4]
```diff
@@ -250,6 +250,44 @@ impl<B: GfxBackend> Adapter<B> {
         }
     }
 
+    pub fn get_swap_chain_preferred_format(
```
Comment (kvark) [inline wgpu-core/src/instance.rs:253]: we'd also need a method on the `Global` object that takes an `AdapterId` and `SurfaceId` and returns the same result

File: wgpu-core/src/instance.rs (hunk 3 above)
Comment (niklaskorz) [reply wgpu-core/src/instance.rs:255]: Not necessarily, but I only found the `get_surface_mut` function for getting the gfx-hal surface of a wgpu surface. If there's an immutable alternative, I'd be glad to use it.

Comment (kvark) [reply wgpu-core/src/instance.rs:255]: that function doesn't force you to accept `&mut` though, you can still have a `&` argument

Comment (niklaskorz) [reply wgpu-core/src/instance.rs:255]: Not sure if I'm missing something, but get_surface_mut only appears to take &mut as input. How would I take an immutable reference and pass it as a mutable reference if the type is not copyable?

Comment (kvark) [reply wgpu-core/src/instance.rs:255]: oh, sorry, I see now :)

**Review by kvark — APPROVED** [review]

**Comment by kvark** [comment]
Please squash and make this a non-draft. I think we are good to go!

**Comment by niklaskorz** [comment]
Checking first if it actually works as intended on Android Vulkan, should be ready in a moment :)

**Comment by kvark** [comment]
thank you!
bors r+

**Pull request merged**